---
title: "lixinxin_lab3"
author: "lixinxin"
date: "2019/4/3"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# 2 Load workflow environment
## 2.1 generate workflow
```{r}
library(systemPipeR)
library(systemPipeRdata)
genWorkenvir(workflow="chipseq")
setwd("chipseq")
```

## 2.2 Experiment de nition provided by targets file
```{r}
targetspath <- system.file("extdata", "targets_chip.txt", package="systemPipeR")
targets <- read.delim(targetspath, comment.char = "#")
targets[1:4,-c(5,6)]
```

# 3 Read preprocessing
## 3.1 Read quality ltering and trimming
```{r}
setwd("chipseq")
args <- systemArgs(sysma="param/trim.param", mytargets="targets_chip.txt")
outpaths(args)[1:4]; sysargs(args)[1:4]

filterFct <- function(fq, cutoff=20, Nexceptions=0)
{
    qcount <- rowSums(as(quality(fq), "matrix") <= cutoff)
    fq[qcount <= Nexceptions] # Retains reads where Phred scores are >= cutoff with N exceptions
}
preprocessReads(args=args, Fct="filterFct(fq, cutoff=20, Nexceptions=0)", batchsize=100000)
writeTargetsout(x=args, file="targets_chip_trim.txt", overwrite=TRUE)
```


## 3.2 FASTQ quality report
```{r}
setwd("chipseq")
args <- systemArgs(sysma="param/bowtieSE.param", mytargets="targets_chip_trim.txt")
outpaths(args)[1:4]; sysargs(args)[1:4]
fqlist <- seeFastq(fastq=infile1(args), batchsize=100000, klength=8)
pdf("./results/fastqReport.pdf", height=18, width=4*length(fqlist))
seeFastqPlot(fqlist)
dev.off()
```


# 4 Alignments
## 4.1 Read mapping with Bowtie2
```{r}
setwd("chipseq")
args <- systemArgs(sysma="param/bowtieSE.param", mytargets="targets_chip_trim.txt")
outpaths(args)[1:4]; sysargs(args)[1:4]
moduleload(modules(args)) # Skip if a module system is not used
system("bowtie2-build ./data/tair10.fasta ./data/tair10.fasta") # Indexes reference genome
runCommandline(args)
writeTargetsout(x=args, file="targets_bam.txt", overwrite=TRUE)
file.exists(outpaths(args))
```

## 4.2 Read and alignment stats
```{r}
setwd("chipseq")
read_statsDF <- alignStats(args=args)
write.table(read_statsDF, "results/alignStats.xls", row.names=FALSE, quote=FALSE, sep="nt")
read.delim("results/alignStats.xls")
```


## 4.3 Create symbolic links for viewing BAM les in IGV
## run in linux platmform !!!!!!!!!
```{r}
#setwd("chipseq")
#symLink2bam(sysargs=args,
#            htmldir=c("~/.html/", "somedir/"),
#            urlbase="http://biocluster.ucr.edu/~tgirke/",
#            urlfile="./results/IGVurl.txt"
#            )
```


# 5 Peak calling with MACS2
## 5.1 Merge BAM les of replicates prior to peak calling
```{r}
setwd("chipseq")
args <- systemArgs(sysma=NULL, mytargets="targets_bam.txt")
outpaths(args)[1:4]; sysargs(args)[1:4]
args_merge <- mergeBamByFactor(args, overwrite=TRUE)
outpaths(args_merge)[1:4]; sysargs(args_merge)[1:4]
writeTargetsout(x=args_merge, file="targets_mergeBamByFactor.txt", overwrite=TRUE)
```

## 5.2 Peak calling without input/reference sample
```{r}
setwd("chipseq")
args <- systemArgs(sysma="param/macs2_noinput.param", mytargets="targets_mergeBamByFactor.txt")
outpaths(args)[1:4]; sysargs(args)[1:4]
runCommandline(args)
file.exists(outpaths(args))
writeTargetsout(x=args, file="targets_macs.txt", overwrite=TRUE)
```


## 5.3 Peak calling with input/reference sample
```{r}
setwd("chipseq")
writeTargetsRef(infile="targets_mergeBamByFactor.txt",
                outfile="targets_bam_ref.txt",
                silent=FALSE,
                overwrite=TRUE
                )
args <- systemArgs(sysma="param/macs2.param", mytargets="targets_bam_ref.txt")
outpaths(args)[1:4]; sysargs(args)[1:4]
runCommandline(args)
file.exists(outpaths(args))
writeTargetsout(x=args, file="targets_macs.txt", overwrite=TRUE)
```

# 6 Annotate peaks with genomic context
## 6.1 Annotation with ChIPpeakAnno package
```{r}
setwd("chipseq")
library(ChIPpeakAnno); library(GenomicFeatures)
args <- systemArgs(sysma="param/annotate_peaks.param", mytargets="targets_macs.txt")
outpaths(args)[1:4]; sysargs(args)[1:4]
txdb <- loadDb("./data/tair10.sqlite")
ge <- genes(txdb, columns=c("tx_name", "gene_id"))
for(i in seq(along=args))
{
    peaksGR <- as(read.delim(infile1(args)[i], comment="#"), "GRanges")
    annotatedPeak <- annotatePeakInBatch(peaksGR, AnnotationData=genes(txdb))
    df <- data.frame(as.data.frame(annotatedPeak), as.data.frame(values(ge[values(annotatedPeak)$feature,])))
    write.table(df, outpaths(args[i]), quote=FALSE, row.names=FALSE, sep="\t")
}
writeTargetsout(x=args, file="targets_peakanno.txt", overwrite=TRUE)
```

# 7 Count reads overlapping peak regions
```{r}
setwd("chipseq")
library(GenomicRanges)
args <- systemArgs(sysma="param/count_rangesets.param", mytargets="targets_macs.txt")
outpaths(args)[1:4]; sysargs(args)[1:4]
args_bam <- systemArgs(sysma=NULL, mytargets="targets_bam.txt")
outpaths(args_merge)[1:4]; sysargs(args_merge)[1:4]
bfl <- BamFileList(outpaths(args_bam), yieldSize=50000, index=character())
countDFnames <- countRangeset(bfl,
                              args,
                              mode="Union",
                              ignore.strand=TRUE
                              )
writeTargetsout(x=args, file="targets_countDF.txt", overwrite=TRUE)
```

# 8 Di erential binding analysis of peaks
```{r}
setwd("chipseq")
args_diff <- systemArgs(sysma="param/rundiff.param", mytargets="targets_countDF.txt")
outpaths(args_diff)[1:4]; sysargs(args_diff)[1:4]
cmp <- readComp(file=args_bam, format="matrix")
dbrlist <- runDiff(args=args_diff,
                   diffFct=run_edgeR,
                   targets=targetsin(args_bam),
                   cmp=cmp[[1]],
                   independent=TRUE,
                   dbrfilter=c(Fold=2, FDR=1)
                   )
writeTargetsout(x=args_diff, file="targets_rundiff.txt", overwrite=TRUE)
```

# 9 GO term enrichment analysis
```{r}
setwd("chipseq")
args <- systemArgs(sysma="param/macs2.param", mytargets="targets_bam_ref.txt")
outpaths(args)[1:4]; sysargs(args)[1:4]
args_anno <- systemArgs(sysma="param/annotate_peaks.param", mytargets="targets_macs.txt")
outpaths(args_anno)[1:4]; sysargs(args_anno)[1:4]
annofiles <- outpaths(args_anno)
gene_ids <- sapply(names(annofiles), function(x) unique(as.character(read.delim(annofiles[x])[,"gene_id"])))
load("data/GO/catdb.RData")
BatchResult <- GOCluster_Report(catdb=catdb,
                                setlist=gene_ids,
                                method="all",
                                id_type="gene",
                                CLSZ=2,
                                cutoff=0.001
                                )
```

# 10 Motif analysis
## 10.1 Parse DNA sequences of peak regions from genome
```{r}
setwd("chipseq")
library(Biostrings); library(seqLogo); library(BCRANK)
args <- systemArgs(sysma="param/annotate_peaks.param", mytargets="targets_macs.txt")
outpaths(args)[1:4]; sysargs(args)[1:4]
rangefiles <- infile1(args)
for(i in seq(along=rangefiles))
{
    df <- read.delim(rangefiles[i], comment="#")
    peaks <- as(df, "GRanges")
    names(peaks) <- paste0(as.character(seqnames(peaks)), "_", start(peaks), "-", end(peaks))
    peaks <- peaks[order(values(peaks)$X.log10.pvalue., decreasing=TRUE)]
    pseq <- getSeq(FaFile("./data/tair10.fasta"), peaks)
    names(pseq) <- names(peaks)
    writeXStringSet(pseq, paste0(rangefiles[i], ".fasta"))
}
```

## 10.2 Motif discovery with BCRANK
```{r}
setwd("chipseq")
set.seed(0)
BCRANKout <- bcrank(paste0(rangefiles[1], ".fasta"), restarts=25, use.P1=TRUE, use.P2=TRUE)
toptable(BCRANKout)
topMotif <- toptable(BCRANKout, 1)
weightMatrix <- pwm(topMotif, normalize = FALSE)
weightMatrixNormalized <- pwm(topMotif, normalize = TRUE)
pdf("results/seqlogo.pdf")
seqLogo(weightMatrixNormalized)
dev.off()
```
warnings()
