---
title: "zhongquan_lab4"
author: "ZhongQuan"
date: "2019/4/13"
output:
  word_document: default
  html_document: default
---

```{r}
# 1 Introduction


# 2 Sample de nitions and environment settings
## 2.1 Environment settings and input data

## 2.2 Required packages and resources
library(systemPipeR)
library(systemPipeRdata)
genWorkenvir(workflow="rnaseq")
setwd("rnaseq")


## 2.3 Experiment de nition provided by targets file
targetspath <- system.file("extdata", "targets.txt", package="systemPipeR")
targets <- read.delim(targetspath, comment.char = "#")[,1:4]
targets

# 3 Read preprocessing
## 3.1 FASTQ quality report
args <- systemArgs(sysma="param/tophat.param", mytargets="targets.txt")
# run in linux system.
fqlist <- seeFastq(fastq=infile1(args), batchsize=100000, klength=8)
seeFastqPlot(fqlist)

# 4 Alignments
## 4.1 Read mapping with Bowtie2/Tophat2
args <- systemArgs(sysma="param/tophat.param", mytargets="targets.txt")
sysargs(args)[1] # Command-line parameters for first FASTQ file

# 
system("bowtie2-build ./data/tair10.fasta ./data/tair10.fasta")
runCommandline(args)

# skip this part
# using computer cluster
# moduleload(modules(args))
# system("bowtie2-build ./data/tair10.fasta ./data/tair10.fasta")
# resources <- list(walltime="20:00:00", nodes=paste0("1:ppn=", cores(args)), memory="10gb")
# reg <- clusterRun(args, conffile=".BatchJobs.R", template="param/torque.tmpl", Njobs=18, runid="01", resourceList=resources)
# waitForJobs(reg)

# check if all bam file have been created
file.exists(outpaths(args))

## 4.2 Read and alignment stats
read_statsDF <- alignStats(args=args)
write.table(read_statsDF, "results/alignStats.xls", row.names=FALSE, quote=FALSE, sep="\t")
read.table(system.file("extdata", "alignStats.xls", package="systemPipeR"), header=TRUE)[1:4,]

## skip this part
## 4.3 Create symbolic links for viewing BAM les in IGV
## run in linux system
# symLink2bam(sysargs=args, htmldir=c("~/.html/", "somedir/"),
#             urlbase="http://biocluster.ucr.edu/~tgirke/",
#             urlfile="./results/IGVurl.txt")

# 5 Read quanti cation per annotation range
## 5.1 Read counting with summarizeOverlaps in parallel mode using multiple cores
library("GenomicFeatures"); library(BiocParallel)
txdb <- loadDb("./data/tair10.sqlite")
eByg <- exonsBy(txdb, by=c("gene"))
bfl <- BamFileList(outpaths(args), yieldSize=50000, index=character())
multicoreParam <- MulticoreParam(workers=8); register(multicoreParam); registered()
counteByg <- bplapply(bfl, function(x) summarizeOverlaps(eByg, x, mode="Union",ignore.strand=TRUE,inter.feature=FALSE,singleEnd=TRUE))
countDFeByg <- sapply(seq(along=counteByg), function(x) assays(counteByg[[x]])$counts)
rownames(countDFeByg) <- names(rowRanges(counteByg[[1]])); colnames(countDFeByg) <- names(bfl)
rpkmDFeByg <- apply(countDFeByg, 2, function(x) returnRPKM(counts=x, ranges=eByg))
write.table(countDFeByg, "results/countDFeByg.xls", col.names=NA, quote=FALSE, sep="\t")
write.table(rpkmDFeByg, "results/rpkmDFeByg.xls", col.names=NA, quote=FALSE, sep="\t")
read.delim("results/countDFeByg.xls", row.names=1, check.names=FALSE)[1:4,1:5]
read.delim("results/rpkmDFeByg.xls", row.names=1, check.names=FALSE)[1:4,1:4]


## 5.2 Sample-wise correlation analysis
library(DESeq2, quietly=TRUE); library(ape, warn.conflicts=FALSE)
countDF <- as.matrix(read.table("./results/countDFeByg.xls"))
colData <- data.frame(row.names=targetsin(args)$SampleName, condition=targetsin(args)$Factor)
dds <- DESeqDataSetFromMatrix(countData = countDF, colData = colData, design = ~ condition)
d <- cor(assay(rlog(dds)), method="spearman")
hc <- hclust(dist(1-d))
plot.phylo(as.phylo(hc), type="p", edge.col="blue", edge.width=2, show.node.label=TRUE, no.margin=TRUE)


# 6 Analysis of di erentially expressed genes with edgeR
library(edgeR)
countDF <- read.delim("./results/countDFeByg.xls", row.names=1, check.names=FALSE)
targets <- read.delim("targets.txt", comment="#")
cmp <- readComp(file="targets.txt", format="matrix", delim="-")
edgeDF <- run_edgeR(countDF=countDF, targets=targets, cmp=cmp[[1]], independent=FALSE, mdsplot="")

# skip this part
# desc <- read.delim("data/desc.xls")
# desc <- desc[!duplicated(desc[,1]),]
# descv <- as.character(desc[,2]); names(descv) <- as.character(desc[,1])
# edgeDF <- data.frame(edgeDF, Desc=descv[rownames(edgeDF)], check.names=FALSE)


write.table(edgeDF, "./results/edgeRglm_allcomp.xls", quote=FALSE, sep="\t", col.names = NA)
edgeDF <- read.delim("results/edgeRglm_allcomp.xls", row.names=1, check.names=FALSE)

DEG_list <- filterDEGs(degDF=edgeDF, filter=c(Fold=1, FDR=20), plot=TRUE)

write.table(DEG_list$Summary, "./results/DEGcounts.xls", quote=FALSE, sep="\t", row.names=FALSE)


vennsetup <- overLapper(DEG_list$Up[6:9], type="vennsets")
vennsetdown <- overLapper(DEG_list$Down[6:9], type="vennsets")

vennPlot(list(vennsetup, vennsetdown), mymain="", mysub="", colmode=2, ccol=c("blue", "red"))




# 6.1 GO term enrichment analysis of DEGs
## 6.1.1 Obtain gene-to-GO mappings
library("biomaRt")
listMarts() # To choose BioMart database
m = useMart(biomart = "plants_mart", host = "plants.ensembl.org")
listDatasets(m)
m <- useMart("plants_mart", dataset="athaliana_eg_gene", host = "plants.ensembl.org")
listAttributes(m) # Choose data types you want to download
go <- getBM(attributes=c("go_id", "tair_locus", "namespace_1003"), mart=m)
go <- go[go[,3]!="",]; go[,3] <- as.character(go[,3])
go[go[,3]=="molecular_function", 3] <- "F"; go[go[,3]=="biological_process", 3] <- "P"; go[go[,3]=="cellular_component", 3] <- "C"
go[1:4,]
dir.create("./data/GO")
write.table(go, "data/GO/GOannotationsBiomart_mod.txt", quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
catdb <- makeCATdb(myfile="data/GO/GOannotationsBiomart_mod.txt", lib=NULL, org="", colno=c(1,2,3))
save(catdb, file="data/GO/catdb.RData")


## 6.1.2 Batch GO term enrichment analysis
load("data/GO/catdb.RData")
DEG_list <- filterDEGs(degDF=edgeDF, filter=c(Fold=2, FDR=50), plot=FALSE)
up_down <- DEG_list$UporDown; names(up_down) <- paste(names(up_down), "_up_down", sep="")
up <- DEG_list$Up; names(up) <- paste(names(up), "_up", sep="")
down <- DEG_list$Down; names(down) <- paste(names(down), "_down", sep="")
DEGlist <- c(up_down, up, down)
DEGlist <- DEGlist[sapply(DEGlist, length) > 0]
BatchResult <- GOCluster_Report(catdb=catdb, setlist=DEGlist, method="all", id_type="gene", CLSZ=2, cutoff=0.01)
goslimvec <- as.character(getBM(attributes=c("goslim_goa_accession"), mart=m)[,1])
BatchResultslim <- GOCluster_Report(catdb=catdb, setlist=DEGlist, method="slim", id_type="gene", myslimv=goslimvec)

## 6.1.3 Plot batch GO term results
gos <- BatchResultslim[grep("M6-V6_up_down", BatchResultslim$CLID), ]
gos <- BatchResultslim
 goBarplot(gos, gocat="MF")
goBarplot(gos, gocat="BP")
goBarplot(gos, gocat="CC")

# 7 Clustering and heat maps
library(pheatmap)
geneids <- unique(as.character(unlist(DEG_list[[1]])))
y <- assay(rlog(dds))[geneids, ]

pheatmap(y, scale="row", clustering_distance_rows="correlation", clustering_distance_cols="correlation")

```
